Userland/Libraries/LibJS/Runtime/NativeFunction.cpp

```diff
@@ -219,7 +219,7 @@ ThrowCompletionOr<Object*> NativeFunction::internal_construct(MarkedVector<Value
     vm.pop_execution_context();
 
     // 12. Return ? result.
-    return result;
+    return *TRY(result);
```

Comment:
Alright so this is a bit of a journey. This is the line that is tripping up test-js.

If this line changes to:
```c++
    // 12. Return ? result.
    dbgln("!!! {}", result.is_error());
    auto result2 = result;
    dbgln("    {} {}", result.is_error(), result2.is_error());
    return *TRY(result2);
```

At the time of crash, this prints:
```
!!! false
    false true
```

So the copy constructor of `ThrowCompletionOr` is turning this into a throw completion, which is what happens when `TRY(result)` is expanded to `auto _temporary_result = (result)`. 

That copy constructor initializes its `Optional<Completion> m_throw_completion` member via the `Optional<Completion>` copy constructor. `Optional<Completion>` is specialized for some reason, and has a plain `JS::Completion m_value` member. The copy constructor for this thing is:

```c++
    Optional(Optional<JS::Completion> const& other)
    {
        if (other.has_value())
            m_value = other.m_value;
    }
```

Here, `other.has_value()` is false, but that doesn't matter. The first thing that happens is `m_value` is default constructed, which is:

```c++
    ALWAYS_INLINE Completion()
        : Completion(js_undefined())
    {
    }
```

So `m_value` is undefined. This specialization's `is_empty` just checks if `m_value` is empty, rather than undefined. So when `ThrowCompletionOr::is_error` checks if `m_throw_completion.is_empty()`, it thinks it has an error, because the value is not empty.

Then we crash in the `TRY()` in NativeFunction because we return undefined as an error, which isn't the case.

This is fixed by making sure the `Optional<Completion>` specialization is initialized to empty:
```diff
diff --git a/Userland/Libraries/LibJS/Runtime/Completion.h b/Userland/Libraries/LibJS/Runtime/Completion.h
index 0367d25727..1ca4fa4260 100644
--- a/Userland/Libraries/LibJS/Runtime/Completion.h
+++ b/Userland/Libraries/LibJS/Runtime/Completion.h
@@ -131,10 +131,7 @@ class Optional<JS::Completion> {
 public:
     using ValueType = JS::Completion;
 
-    Optional()
-        : m_value(JS::Completion(JS::Completion::EmptyTag {}))
-    {
-    }
+    Optional() = default;
 
     Optional(Optional<JS::Completion> const& other)
     {
@@ -228,7 +225,7 @@ public:
     JS::Completion* operator->() { return &value(); }
 
 private:
-    JS::Completion m_value;
+    JS::Completion m_value { JS::Completion::EmptyTag {} };
 };
 
 }
```

*How* this hasn't happened before is beyond me but :vomiting_face:

Reply:
By the way this all could go away after the next commit to, because once `construct` returns `ThrowCompletionOr<NonnullGCPtr<Object*>>`, this return can be reverted back to:
```c++
return result;
```

which would have hidden this issue even longer :upside_down_face:

Reply:
Great investigation, thanks! Added a commit to fix this with attribution :)